wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,709 @@
+| `SparseCumlArray` | `@api_base_return_sparse_array(get_output_type=True)` | Functions can return any sparse array-like object |
+| `dict`, `tuple`, `list` or `typing.Union` | `@api_base_return_generic(get_output_type=True)` | Functions must return a generic object that contains an array-like object. No sparse arrays are supported |
+
+### Examples
+
+**Before: `predict()`**
+```python
+@with_cupy_rmm
+def predict(self, X, y):
+   # Determine the output type and dtype
+   out_type = self._get_output_type(y)
+
+   # Convert to CumlArray
+   X_m = input_to_cuml_array(X, order="K").array
+
+   # Call a cuda function
+   someCudaFunction(X_m.ptr)
+
+   # Convert the CudaArray to the desired output
+   return X_m.to_output(output_type=out_type)
+```
+
+**After: `predict()`**
+```python
+def predict(self, X) -> CumlArray:
+   # Convert to CumlArray
+   X_m = input_to_cuml_array(X, order="K").array
+
+   # Call a cuda function
+   someCudaFunction(X_m.ptr)
+
+   # Convert the CudaArray to the desired output
+   return X_m
+```
+
+**Before: `fit()`**
+
+```python
+@with_cupy_rmm
+def fit(self, X):
+   # Set the base input attributes
+   self._set_base_attributes(output_type=X, n_features=X)
+
+   self.coef_ = input_to_cuml_array(X, order="K").array
+
+   return self
+```
+
+**After: `fit()`**
+
+```python
+def fit(self, X) -> "KMeans:
+   self.coef_ = input_to_cuml_array(X, order="K").array
+
+   return self
+```
+
+## Do's And Don'ts
+
+### **Do:** Add Typing Information to Estimator Functions
+
+Adding the return type to estimator functions will allow the `Base` meta-class to automatically decorate functions based on their return type.
+
+**Do this:**
+```python
+def fit(self, X, y, convert_dtype=True) -> "KNeighborsRegressor":
+def predict(self, X, convert_dtype=True) -> CumlArray:
+def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity') -> SparseCumlArray:
+def predict(self, start=0, end=None, level=None) -> typing.Union[CumlArray, float]:
+```
+
+**Not this:**
+```python
+def fit(self, X, y, convert_dtype=True):
+def predict(self, X, convert_dtype=True) :
+def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity'):
+def predict(self, start=0, end=None, level=None):
+```
+
+### **Do:** Return Array-Like Objects Directly
+
+There is no need to convert the array type before returning it. Simply return any array-like object and it will be automatically converted
+
+**Do this:**
+```python
+def predict(self, X, y) -> CumlArray:
+   np_arr = np.ones((10,))
+
+   return np_arr
+```
+
+**Not this:**
+```python
+def predict(self, X, y) -> CumlArray:
+   np_arr = np.ones((10,))
+
+   cuml_arr = input_to_cuml_array(X, order="K").array
+
+   return cuml_arr.to_output(self._get_output_type(X))
+```
+
+### Example Decorator Use Cases
+
+The following section shows some before and after examples of how the decorators should be used with default options.
+
+**Example 1: 
```

Comment:
Doesn't seem to format correctly, maybe *Example 1:*?

Reply:
Fixed. Resolving comment